File: backend/src/prechecks/concrete/cluster/no-relocating-shards.precheck.ts
import { ElasticClient } from "../../../clients/elastic.client";
import { BaseClusterPrecheck } from "../../base/base-cluster-precheck";
import { ExecutionMode, PrecheckType } from "../../types/enums";
import { PrecheckExecutionRequest, ClusterContext } from "../../types/interfaces";

export class NoRelocatingShardsPrecheck extends BaseClusterPrecheck {
	constructor() {
		super({
			id: "elasticsearch_no_relocating_shards",
			name: "No relocating shards",
			type: PrecheckType.CLUSTER,
			mode: ExecutionMode.CODE,
		});
	}

	protected async runForContext(request: PrecheckExecutionRequest<ClusterContext>): Promise<void> {
		const client = await ElasticClient.buildClient(request.cluster.clusterId);
		const es = client.getClient();

		const health = await es.cluster.health();
		const relocatingCount = health.relocating_shards;

		this.addLog(request, `Relocating shards count: ${relocatingCount}. Expected: 0.`, "");

		if (relocatingCount > 0) {
			const allShards = await es.cat.shards({ format: "json" });
			const relocatingShards = allShards.filter((s) => s.state === "RELOCATING");

			for (const shard of relocatingShards) {
				await this.addLog(
					request,
					`Relocating shard: index=${shard.index}, shard=${shard.shard}, from=${shard.node}`
				);
			}

			throw new Error(`Relocating shards check failed. ${relocatingCount} shard(s) are currently relocating.`);
		}
	}
}

File: server/src/main/java/co/hyperflex/prechecks/concrete/cluster/NoRelocatingShardsPrecheck.java
package co.hyperflex.prechecks.concrete.cluster;

import co.elastic.clients.elasticsearch.ElasticsearchClient;
import co.elastic.clients.elasticsearch.cat.shards.ShardsRecord;
import co.hyperflex.prechecks.contexts.ClusterContext;
import co.hyperflex.prechecks.core.BaseClusterPrecheck;
import co.hyperflex.prechecks.core.PrecheckLogger;
import java.io.IOException;
import java.util.List;
import org.springframework.stereotype.Component;

@Component
public class NoRelocatingShardsPrecheck extends BaseClusterPrecheck {

  @Override
  public String getName() {
    return "No relocating shards";
  }

  @Override
  public void run(ClusterContext context) {
    ElasticsearchClient client = context.getElasticClient().getElasticsearchClient();
    PrecheckLogger logger = context.getLogger();

    try {

      List<ShardsRecord> shards =
          client.cat().shards(s -> s.h("index", "shard", "state", "node")).valueBody();

      List<ShardsRecord> relocatingShards =
          shards.stream().filter(s -> "RELOCATING".equalsIgnoreCase(s.state())).toList();

      for (ShardsRecord shard : relocatingShards) {
        logger.error("Relocating shard: index=%s, shard=%s, from=%s", shard.index(), shard.shard(),
            shard.node());
      }

      if (!relocatingShards.isEmpty()) {
        throw new RuntimeException(
            String.format("Relocating shards check failed. %d shard(s) are currently relocating.",
                relocatingShards.size()));
      }

    } catch (IOException e) {
      logger.error("Failed to check relocating shards", e);
      throw new RuntimeException("Failed to check relocating shards", e);
    }
  }
}
